presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotQueryGenerator.java

```diff
@@ -516,5 +518,33 @@ public PinotQueryGeneratorContext visitTopN(TopNNode node, PinotQueryGeneratorCo
             checkSupported(node.getStep().equals(TopNNode.Step.SINGLE), "Can only push single logical topn in");
             return context.withTopN(getOrderingScheme(node), node.getCount()).withOutputColumns(node.getOutputVariables());
         }
+
+        @Override
+        public PinotQueryGeneratorContext visitDistinctLimit(DistinctLimitNode node, PinotQueryGeneratorContext context)
+        {
+            context = node.getSource().accept(this, context);
+            requireNonNull(context, "context is null");
+            checkSupported(!forbidBrokerQueries, "Cannot push distinctLimit in segment mode");
+            LinkedHashSet<VariableReferenceExpression> groupByColumns = new LinkedHashSet<>(node.getDistinctVariables());
+            if (!useSqlSyntax) {
+                // Handling PQL by adding hidden expression: count(*)
+                // E.g. `SELECT DISTINCT A, B FROM myTable LIMIT 10`
+                // In Pql mode, the generated pql is `SELECT count(*) FROM myTable GROUP BY A, B LIMIT 10`.
+                checkSupported(!context.hasAggregation(), "Aggregation already exists. Pinot doesn't support DistinctLimit with existing Aggregation");
+                checkSupported(!context.hasGroupBy(), "GroupBy already exists. Pinot doesn't support DistinctLimit with existing GroupBy");
+                VariableReferenceExpression hidden = new VariableReferenceExpression(UUID.randomUUID().toString(), BigintType.BIGINT);
+                Map<VariableReferenceExpression, Selection> newSelections = new HashMap<>(context.getSelections());
+                newSelections.put(hidden, new Selection("count(*)", DERIVED));
+                LinkedHashSet<VariableReferenceExpression> newOutputs = new LinkedHashSet<>(groupByColumns);
+                newOutputs.add(hidden);
```

Comment:
Correct me if I am wrong: But is this block of code duplicated ? If so, can we DRY it ?

Reply:
extract to a method to set hiden column